api/metric/api.go

```diff
@@ -16,66 +16,97 @@ package metric
 
 import (
 	"context"
+	"time"
 
 	"go.opentelemetry.io/api/core"
 	"go.opentelemetry.io/api/registry"
 	"go.opentelemetry.io/api/unit"
 )
 
-type MetricType int
+// Kind categorizes different kinds of metric.
+type Kind int
 
+//go:generate stringer -type=Kind
 const (
-	Invalid    MetricType = iota
-	Gauge                 // Supports Set()
-	Cumulative            // Supports Inc()
+	Invalid        Kind = iota
+	CumulativeKind      // Supports Add()
+	GaugeKind           // Supports Set()
+	MeasureKind         // Supports Record()
 )
 
-type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Recorder interface {
+	// RecordAt allows the SDK to observe a single metric event
+	RecordAt(ctx context.Context, t time.Time, value float64)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+// Meter is an interface to the metrics portion of the OpenTelemetry SDK.
+type Meter interface {
+	// RecorderFor returns a handle for observing single measurements.
+	RecorderFor(context.Context, Instrument, ...core.KeyValue) Recorder
```

Comment:
In my APIs I invert this.  `ObserverAt(tsid, meta)`

Reply:
I'm not sure I understand. I think `tsid` corresponds with `Instrument` and `meta` corresponds with `...core.KeyValue`.  Do you prefer "At" to "For"? I stuck with "Record" as a stem because of the existing verbs in the spec, which talk about recording metrics. I could be convinced that "Observer" is as good.

Reply:
In the past I've broken time series data into 4 components

1. Time
2. Value
3. Time series identifier (tsid)
4. Metadata.

Time and value are obvious.  The TSID is the unique identifier that metrics aggregate upon (for example, for SignalFx or Datadog or cloudwatch this is a metric name and dimensions/tags). Metadata are extra concepts that don't change aggregation.  Those are description/unit/etc.

I usually clearly separate those 4 into their own components.  I think opentelemetry tries to combine tsid and metadata into the same struct.  It may not be possible to split (3) and (4) in a generic way.